// vnpay/lib/vnpay.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.VNPay = void 0;
const crypto_1 = __importDefault(require("crypto"));
const moment_timezone_1 = __importDefault(require("moment-timezone"));
const class_validator_1 = require("class-validator");
const constants_1 = require("./constants");
const dtos_1 = require("./dtos");
const enums_1 = require("./enums");
const common_1 = require("./utils/common");
const return_query_from_vnpay_dto_1 = require("./dtos/return-query-from-vnpay.dto");
const verify_return_url_dto_1 = require("./dtos/verify-return-url.dto");
/**
 * VNPay class to support VNPay payment
 * @vi_vn Lớp hỗ trợ thanh toán qua VNPay
 *
 * @example
 * import { VNPay } from 'vnpay';
 *
 * const vnpay = new VNPay({
 *     paymentGateway: 'https://sandbox.vnpayment.vn/paymentv2/vpcpay.html',
 *     tmnCode: 'TMNCODE',
 *     secureSecret: 'SERCRET',
 * });
 *
 * // or setup with async/await
 * const vnpayAsync = await VNPay.setup({
 *     paymentGateway: 'https://sandbox.vnpayment.vn/paymentv2/vpcpay.html',
 *     tmnCode: 'TMNCODE',
 *     secureSecret: 'SERCRET',
 * });
 *
 * const tnx = '12345678';
 * const urlString = await vnpay.buildPaymentUrl({
 *     vnp_Amount: 100000,
 *      vnp_IpAddr: '192.168.0.1',
 *      vnp_ReturnUrl: 'http://localhost:8888/order/vnpay_return',
 *      vnp_TxnRef: tnx,
 *      vnp_OrderInfo: `Thanh toan cho ma GD: ${tnx}`,
 * }),
 */
class VNPay {
    constructor(_a) {
        var { paymentGateway = constants_1.PAYMENT_GATEWAY_SANDBOX } = _a, init = __rest(_a, ["paymentGateway"]);
        this.CRYPTO_ALGORITHM = 'sha512';
        this.CRYPTO_ENCODING = 'utf-8';
        this.vnp_Version = constants_1.VNP_VERSION;
        this.vnp_Command = constants_1.VNP_DEFAULT_COMMAND;
        this.vnp_CurrCode = constants_1.CURR_CODE_VND;
        this.vnp_Locale = enums_1.VnpLocale.VN;
        this.vnp_OrderType = enums_1.VnpOrderType.OTHER;
        this.globalConfig = new dtos_1.ConfigVnpayDTO(Object.assign({ paymentGateway }, init));
    }
    /**
     * Setup the VNPay config
     * @vi_vn Phương thức thiết lập cấu hình cho VNPay
     *
     * @param {ConfigVnpayDTO} init  - The config to setup
     * @returns {VNPay} The VNPay instance
     */
    static setup(init) {
        return __awaiter(this, void 0, void 0, function* () {
            const initValid = new dtos_1.ConfigVnpayDTO(init);
            const err = yield (0, class_validator_1.validate)(initValid);
            if (err.length > 0) {
                throw err;
            }
            return new this(initValid);
        });
    }
    /**
     * Default config for VNPay
     * @vi_vn Cấu hình mặc định cho VNPay
     */
    get configDefault() {
        return {
            vnp_Version: this.vnp_Version,
            vnp_Command: this.vnp_Command,
            vnp_CurrCode: this.vnp_CurrCode,
            vnp_Locale: this.vnp_Locale,
            vnp_OrderType: this.vnp_OrderType,
        };
    }
    /**
     * Set the default config for VNPay
     * @vi_vn Phương thức thiết lập cấu hình mặc định cho VNPay
     */
    set configDefault(config) {
        var _a, _b, _c, _d, _e;
        this.vnp_Version = (_a = config.vnp_Version) !== null && _a !== void 0 ? _a : this.vnp_Version;
        this.vnp_Command = (_b = config.vnp_Command) !== null && _b !== void 0 ? _b : this.vnp_Command;
        this.vnp_CurrCode = (_c = config.vnp_CurrCode) !== null && _c !== void 0 ? _c : this.vnp_CurrCode;
        this.vnp_Locale = (_d = config.vnp_Locale) !== null && _d !== void 0 ? _d : this.vnp_Locale;
        this.vnp_OrderType = (_e = config.vnp_OrderType) !== null && _e !== void 0 ? _e : this.vnp_OrderType;
    }
    validateGlobalConfig() {
        if (!this.globalConfig.secureSecret) {
            return new Error('Missing secure secret');
        }
        if (!this.globalConfig.tmnCode) {
            return new Error('Missing merchant code');
        }
        if (!this.globalConfig.paymentGateway) {
            return new Error('Missing payment gateway');
        }
        return true;
    }
    /**
     * Build the payment url
     * @vi_vn Phương thức xây dựng, tạo thành url thanh toán của VNPay
     *
     * @param {BuildPaymentUrlDTO} payload - Payload that contains the information to build the payment url
     * @returns {string} The payment url string
     */
    buildPaymentUrl(payload) {
        return new Promise((resolve, reject) => {
            const err = this.validateGlobalConfig();
            if (err instanceof Error) {
                return reject(err);
            }
            if (!payload.vnp_ReturnUrl) {
                payload.vnp_ReturnUrl = this.globalConfig.returnUrl;
            }
            const data = Object.assign(Object.assign({}, this.configDefault), payload);
            const timeGMT7 = (0, moment_timezone_1.default)(new Date()).tz('Asia/Ho_Chi_Minh').format();
            data.vnp_CreateDate = (0, common_1.dateFormat)(new Date(timeGMT7), 'yyyyMMddHHmmss');
            data.vnp_Amount = data.vnp_Amount * 100;
            data.vnp_TmnCode = this.globalConfig.tmnCode;
            const dataValidate = new dtos_1.BuildPaymentUrlDTO(data);
            (0, class_validator_1.validate)(dataValidate).then((err) => {
                if (err.length > 0) {
                    reject(err);
                }
            });
            const redirectUrl = new URL(String(this.globalConfig.paymentGateway));
            Object.entries(data)
                .sort(([key1], [key2]) => key1.toString().localeCompare(key2.toString()))
                .forEach(([key, value]) => {
                // Skip empty value
                if (!value || value === '' || value === undefined || value === null) {
                    return;
                }
                redirectUrl.searchParams.append(key, value.toString());
            });
            const hmac = crypto_1.default.createHmac(this.CRYPTO_ALGORITHM, this.globalConfig.secureSecret);
            const signed = hmac
                .update(Buffer.from(redirectUrl.search.slice(1).toString(), this.CRYPTO_ENCODING))
                .digest('hex');
            redirectUrl.searchParams.append('vnp_SecureHash', signed);
            return resolve(redirectUrl.toString());
        });
    }
    /**
     * Method to verify the return url from VNPay
     * @vi_vn Phương thức xác thực tính đúng đắn của các tham số trả về từ VNPay
     * @param {ReturnQueryFromVNPayDTO} query - The object of data return from VNPay
     * @returns {Promise<VerifyReturnUrlDTO>} The return object
     */
    verifyReturnUrl(query) {
        return new Promise((resolve, reject) => {
            var _a;
            const err = this.validateGlobalConfig();
            if (err !== true) {
                return reject(err);
            }
            const vnpayReturnQuery = new return_query_from_vnpay_dto_1.ReturnQueryFromVNPayDTO(Object.assign({}, query));
            (0, class_validator_1.validate)(vnpayReturnQuery).then((errs) => {
                if (errs.length > 0) {
                    reject(errs);
                }
            });
            const secureHash = vnpayReturnQuery.vnp_SecureHash;
            // Will be remove when append to URLSearchParams
            vnpayReturnQuery.vnp_SecureHash = '';
            vnpayReturnQuery.vnp_SecureHashType = '';
            const outputResults = {
                isSuccess: vnpayReturnQuery.vnp_ResponseCode === '00',
                message: VNPay.getResponseByStatusCode(vnpayReturnQuery.vnp_ResponseCode.toString(), this.configDefault.vnp_Locale),
            };
            const urlReturn = new URL((_a = this.globalConfig.paymentGateway) !== null && _a !== void 0 ? _a : constants_1.PAYMENT_GATEWAY_SANDBOX);
            Object.entries(vnpayReturnQuery)
                .sort(([key1], [key2]) => key1.toString().localeCompare(key2.toString()))
                .forEach(([key, value]) => {
                // Skip empty value
                if (value === '' || value === undefined || value === null) {
                    return;
                }
                urlReturn.searchParams.append(key, value.toString());
            });
            const hmac = crypto_1.default.createHmac(this.CRYPTO_ALGORITHM, this.globalConfig.secureSecret);
            const signed = hmac
                .update(Buffer.from(urlReturn.search.slice(1).toString(), this.CRYPTO_ENCODING))
                .digest('hex');
            if (secureHash === signed) {
                Object.assign(outputResults, {
                    isSuccess: vnpayReturnQuery.vnp_ResponseCode === '00',
                });
            }
            else {
                Object.assign(outputResults, {
                    isSuccess: false,
                    message: 'Wrong checksum',
                });
            }
            const returnObject = new verify_return_url_dto_1.VerifyReturnUrlDTO(Object.assign(Object.assign({}, vnpayReturnQuery), outputResults));
            resolve(returnObject);
        });
    }
    /**
     *
     * @param responseCode The response code from VNPay
     * @param locale The locale to get the response text
     * @returns The response text
     */
    static getResponseByStatusCode(responseCode, locale = enums_1.VnpLocale.VN) {
        var _a;
        const responseMap = new Map([
            ['00', { vn: 'Giao dịch thành công', en: 'Approved' }],
            ['01', { vn: 'Giao dịch đã tồn tại', en: 'Transaction is already exist' }],
            [
                '02',
                {
                    vn: 'Merchant không hợp lệ (kiểm tra lại vnp_TmnCode)',
                    en: 'Invalid merchant (check vnp_TmnCode value)',
                },
            ],
            [
                '03',
                {
                    vn: 'Dữ liệu gửi sang không đúng định dạng',
                    en: 'Sent data is not in the right format',
                },
            ],
            [
                '04',
                {
                    vn: 'Khởi tạo GD không thành công do Website đang bị tạm khoá',
                    en: 'Payment website is not available',
                },
            ],
            [
                '05',
                {
                    vn: 'Giao dịch không thành công do: Quý khách nhập sai mật khẩu thanh toán quá số lần quy định. Xin quý khách vui lòng thực hiện lại giao dịch',
                    en: 'Transaction failed: Too many wrong password input',
                },
            ],
            [
                '06',
                {
                    vn: 'Giao dịch không thành công do Quý khách nhập sai mật khẩu xác thực giao dịch (OTP). Xin quý khách vui lòng thực hiện lại giao dịch.',
                    en: 'Transaction failed: Wrong OTP input',
                },
            ],
            [
                '07',
                {
                    vn: 'Trừ tiền thành công. Giao dịch bị nghi ngờ (liên quan tới lừa đảo, giao dịch bất thường). Đối với giao dịch này cần merchant xác nhận thông qua merchant admin: Từ chối/Đồng ý giao dịch',
                    en: 'This transaction is suspicious',
                },
            ],
            [
                '08',
                {
                    vn: 'Giao dịch không thành công do: Hệ thống Ngân hàng đang bảo trì. Xin quý khách tạm thời không thực hiện giao dịch bằng thẻ/tài khoản của Ngân hàng này.',
                    en: 'Transaction failed: The banking system is under maintenance. Please do not temporarily make transactions by card / account of this Bank.',
                },
            ],
            [
                '09',
                {
                    vn: 'Giao dịch không thành công do: Thẻ/Tài khoản của khách hàng chưa đăng ký dịch vụ InternetBanking tại ngân hàng.',
                    en: 'Transaction failed: Cards / accounts of customer who has not yet registered for Internet Banking service.',
                },
            ],
            [
                '10',
                {
                    vn: 'Giao dịch không thành công do: Khách hàng xác thực thông tin thẻ/tài khoản không đúng quá 3 lần',
                    en: 'Transaction failed: Customer incorrectly validate the card / account information more than 3 times',
                },
            ],
            [
                '11',
                {
                    vn: 'Giao dịch không thành công do: Đã hết hạn chờ thanh toán. Xin quý khách vui lòng thực hiện lại giao dịch.',
                    en: 'Transaction failed: Pending payment is expired. Please try again.',
                },
            ],
            [
                '24',
                {
                    vn: 'Giao dịch không thành công do: Khách hàng hủy giao dịch',
                    en: 'Transaction canceled',
                },
            ],
            [
                '51',
                {
                    vn: 'Giao dịch không thành công do: Tài khoản của quý khách không đủ số dư để thực hiện giao dịch.',
                    en: 'Transaction failed: Your account is not enough balance to make the transaction.',
                },
            ],
            [
                '65',
                {
                    vn: 'Giao dịch không thành công do: Tài khoản của Quý khách đã vượt quá hạn mức giao dịch trong ngày.',
                    en: 'Transaction failed: Your account has exceeded the daily limit.',
                },
            ],
            [
                '75',
                {
                    vn: 'Ngân hàng thanh toán đang bảo trì',
                    en: 'Banking system is under maintenance',
                },
            ],
            ['default', { vn: 'Giao dịch thất bại', en: 'Failure' }],
        ]);
        const respondText = (_a = responseMap.get(responseCode)) !== null && _a !== void 0 ? _a : responseMap.get('default');
        return respondText[locale];
    }
}
exports.VNPay = VNPay;
